### ISSUE TITLE
ascii art renderer

### ISSUE THREAD

**Issue by alixander** [terrastruct/d2#924]
our code is set up well to just plug in an ascii art renderer. i suspect still will be quite hard, though we can treat everything as rectangles to start to make it easier.

**Comment by alixander** [comment]
https://github.com/ironcamel/Graph-Easy

**Comment by i-am-the-slime** [comment]
Any hope for this one?

**Comment by alixander** [comment]
Unfortunately there's no motivating factor right now.

**Comment by AkashJana18** [comment]
/attempt #924 
I would like to work on this issue.
Hey @alixander can you shed some more light how to proceed further?

**Comment by alixander** [comment]
@AkashJana18 since there's a PR open, I'll let him proceed first and ping you if it doesn't work out

**Comment by alixander** [comment]
More context: https://github.com/terrastruct/d2/pull/2353#pullrequestreview-2618517820

**Comment by AkashJana18** [comment]
Hey @DrSensor, would you like to collaborate on this issue, or should I go ahead and start working on it? Or are you planning to handle it solo?

**Comment by alixander** [comment]
I'll give 72 hours to respond to comments each time. I'll keep it moving to ensure people have a fair shot @AkashJana18

**Comment by DrSensor** [comment]
@AkashJana18 i came up with other solution using grid based method so i want to handle it solo. i will mention you if i need help.

**Comment by bo-ku-ra** [comment]
interesting if it is interconvertible.
https://ivanceras.github.io/svgbob-editor/

**Comment by alixander** [comment]
Well we already have SVG renderers so converting ASCII to SVG with another tool doesn't seem useful

**Comment by bo-ku-ra** [comment]
umm, talk about SVG is short sighted...

that is great ascii art.
i believe that d2's ability to output compatibly is to produce beautiful ascii art output.

**Comment by mohamed1408** [comment]
/attempt #924
I would like to work on this issue.
I have one doubt - I have almost completed, but do we need curves in ascii in connections. 
Also shall I work

**Comment by mohamed1408** [comment]
/attempt #924
I would like to work on this issue.
I have one doubt - I have completed basic shapes, but do we need curves in ascii in connections. 
Also shall I work?

eg:
alphabet -> vowels

![Image](https://github.com/user-attachments/assets/4ccf0d4d-a7dd-4ac9-93d9-517d8a99892b)

![Image](https://github.com/user-attachments/assets/b8ebc267-c4e6-4a66-b658-f2d77555e819)

Is this Ok?

**Comment by alixander** [comment]
yes that's fine. can be only non-curve layouts. e.g. should render the landing page example: https://play.d2lang.com/?script=rJJBjtswDEX3OgWBrm2kzU4FeoruC8JibSG0KIj0pMHAdy9kW00yU2A2s6RI_8__6Bcs6uHVAYRv3SDpdxz3EoDxJot1lMaYyAPxZXv-Aj-pzDEhg000EwwSaOtsZReDh_Pp5ABWt7pEdpVy2TUHYgaTK5XmoWjEHI20vQDohJk8qEmh8CugYWvYjamfF7aYmTxYWXbj1e3-BZPO0YyKe6MO3Y_HtgelFD5xZttAEsdEkKUYcsuzRA-vR6SJ_uAoaW0f1GyQiwykKncmJgVHegdkuHFMgcrHNA75O-7-Yeca4dm4_-eYJ0kELKO61blF26GODTIVleQArjHY5OHr-dTmqupx6_7xzDNeCAZkfjf1RKuvmHCo--yGWzi1IhfqAurk4VytMEdQKi8fSYWomfH2Zr4G21NENbdX7TYZR_r-X6ira39x_8ytSt4N3N8AAAD__w%3D%3D&

**Comment by mohamed1408** [comment]
Then it is ok. I will finish my work and let you know.

**Comment by mohamed1408** [comment]
Hi it is taking some time. So I wanna let you know the progress I've made. The pending works are some shapes and connections. I share you the sample (partial) that my code can generate now. Pls let me know if any changes to e done

![Image](https://github.com/user-attachments/assets/1bcb406e-c4bf-4d20-b556-7c6da79a96d3)

[d2_landing_page_example_ascii.txt](https://github.com/user-attachments/files/20156212/d2_landing_page_example_ascii.txt)

FONT: Consolas Regular
FONT SIZE: 12 pt

![Image](https://github.com/user-attachments/assets/4e46e7d9-3ed0-42cc-89c1-b75755b5235b)

**Comment by mohamed1408** [comment]
That's genius. Is this ok with @alixander ?

**Comment by bo-ku-ra** [comment]
the symbols for all SHAPES need to be determined.
also, the design needs to be devised in the hierarchy.

```
e.g.

╭──┬──────╮
│☁ │Cloud Drive │
╰──┴──────╯



╭┤☁├─────╮
│ Cloud Drive   │
│               │
│ ╭─────╮  │
│ │ Default │ │
│ ╰─────╯  │
╰────────╯

```

**Comment by mohamed1408** [comment]
> How about an approach to make all shapes rectangular?
> I think d2 can get smaller results.
> 
> e.g.
> 
> shape: rectangle
> ╭─────╮
> │Default  │
> ╰─────╯
> 
> 
> shape:person
> (^_^)────╮
> │    USER   │
> ╰──────╯
> 
> 
> shape: cloud
>  ☁────╮
> │ CLOUD   │
> ╰─────╯
> 
> ╭┤☁├─────╮
> │  Cloud Drive  │
> ╰────────╯
> 
> ╭──┬──────╮
> │☁ │Cloud Drive │
> ╰──┴──────╯
> 
> 
> 

Hi @alixander can you comment on this?

**Comment by alixander** [comment]
Yes, that is good to me.

**Comment by mohamed1408** [comment]
Hi @alixander it is almost complete. Having small problems in connections arrow head.

Wanted to let you know will be completed in 2 days.
